datafusion/common/src/tree_node.rs

```diff
@@ -289,15 +492,80 @@ pub enum RewriteRecursion {
     Skip,
 }
 
-/// Controls how the [`TreeNode`] recursion should proceed for [`TreeNode::visit`].
+/// Controls how the [`TreeNode`] recursion should proceed for [`TreeNode::visit_down()`] and
+/// [`TreeNode::visit()`].
 #[derive(Debug)]
-pub enum VisitRecursion {
-    /// Continue the visit to this node tree.
+pub enum TreeNodeRecursion {
+    /// Continue the visit to the next node.
     Continue,
-    /// Keep recursive but skip applying op on the children
-    Skip,
-    /// Stop the visit to this node tree.
+
+    /// Prune the current subtree.
+    /// If a preorder visit of a tree node returns [`TreeNodeRecursion::Prune`] then inner
+    /// children and children will not be visited and postorder visit of the node will not
+    /// be invoked.
+    Prune,
```

Comment:
Is this equivalent to `RewriteRecursion::Skip`?  If so, perhaps we can use the same terminology

Reply:
We can keep `Skip` if you prefer. (To me `Prune` better describes that children should not be visited.)